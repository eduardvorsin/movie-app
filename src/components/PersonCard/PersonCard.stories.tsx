import type { Meta, StoryObj } from '@storybook/react';
import PersonCard, { Props } from './PersonCard';
import { useTranslation } from 'react-i18next';

const meta: Meta<typeof PersonCard> = {
	title: 'Layout/PersonCard',
	tags: ['autodocs'],
	argTypes: {
		appearance: {
			description: 'Appearance for the card',
		},
		className: {
			description: 'Additional classes for the PersonCard'
		},
		testId: {
			description: 'Id for testing the component',
		},
		personId: {
			description: 'Id for the link to the page with the person',
		},
		src: {
			description: 'The path to the picture',
		},
		alt: {
			description: 'Alternative text for the image',
		},
		title: {
			description: 'Title for the card',
		},
		showRating: {
			description: 'Allows you to hide/show the rating of a film project',
		},
		rating: {
			description: 'Rating value',
		},
		children: {
			description: 'The text content of the component',
		},
		titleElement: {
			description: 'The html element to be used for the header',
		},
		titleLevel: {
			description: 'Header level',
		},
		onClick: {
			control: false,
			description: 'Allows you to send a callback that is called when you click on the component',
		},
		dictionary: {
			control: false,
			description: 'An object with keys for localization of the component',
		},
		loading: {
			description: 'Allows you to add a lazy loading for an image',
		},
	},
	component: PersonCard,
	parameters: {
		docs: {
			description: {
				component: 'The PersonCard component is a key element of the user interface designed to display information about characters â€“ actors, directors, producers and other participants in the cinematic process.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof PersonCard>;

const PersonCardWithHooks = (props: Omit<Props, 'dictionary'>) => {
	const { t } = useTranslation('common');
	const dictionary = { rating: t('personCard.rating') };

	return (
		<PersonCard
			{...props}
			dictionary={dictionary}
		/>
	);
}

export const Default: Story = {
	args: {
		className: 'max-w-[154px]',
		personId: 9999,
		src: '/assets/images/person-placeholder-l.svg',
		rating: 100,
		alt: '',
		title: 'Title',
		titleLevel: 6,
		children: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => (<PersonCardWithHooks {...args} />),
};

export const ShowRating: Story = {
	args: {
		className: 'max-w-[154px]',
		personId: 9999,
		src: '/assets/images/person-placeholder-l.svg',
		rating: 100,
		showRating: true,
		alt: '',
		title: 'Title',
		titleLevel: 6,
		children: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
	},
	parameters: {
		docs: {
			description: {
				story: 'Allows you to hide or show an item with a rating for a person',
			},
		},
	},
	render: (args) => (<PersonCardWithHooks {...args} />),
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
	},
	args: {
		className: 'max-w-[154px]',
		personId: 9999,
		src: '/assets/images/person-placeholder-l.svg',
		rating: 100,
		alt: '',
		title: 'Title',
		titleLevel: 6,
		children: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
	},
	parameters: {
		docs: {
			description: {
				story: 'The callback that will be called when clicking on the component',
			},
		},
	},
	render: (args) => (<PersonCardWithHooks {...args} />),
};

export const Appearance: Story = {
	args: {
		personId: 9999,
		src: '/assets/images/person-placeholder-l.svg',
		rating: 100,
		alt: '',
		title: 'Title',
		titleLevel: 6,
		children: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
	},
	parameters: {
		docs: {
			description: {
				story: 'The appearance of the card. There are 2 appearance options in total: primary, secondary',
			},
		},
	},
	render: (args) => (
		<div className='flex gap-6 items-start'>
			<PersonCardWithHooks
				{...args}
				className='basis-[154px] shrink-0 grow-0'
				appearance='primary'
			/>
			<PersonCardWithHooks
				{...args}
				className='basis-[154px] shrink-0 grow-0'
				appearance='secondary'
			/>
		</div>
	),
};