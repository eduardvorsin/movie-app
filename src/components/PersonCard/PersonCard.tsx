'use client';
import { MouseEventHandler, ReactNode } from 'react';
import Title from '@/components/Title/Title';
import Link from 'next/link';
import { HeadingElement, HeadingLevel } from '@/types/shared';
import { useParams } from 'next/navigation';
import { Locales, fallbackLng } from '@/i18n/settings';
import { useTranslation } from '@/i18n/client';
import ThemedImage from '../ThemedImage/ThemedImage';
import { routes } from 'src/constants';

type Props = {
	personId: number,
	className?: string,
	testId?: string,
	title?: string,
	titleElement?: HeadingElement,
	titleLevel?: HeadingLevel,
	src: string,
	alt: string,
	showRating?: boolean,
	rating?: number,
	children?: ReactNode,
	onClick?: MouseEventHandler<HTMLAnchorElement>,
};

export default function PersonCard({
	personId,
	className,
	testId,
	title,
	titleElement = 'h6',
	titleLevel = 4,
	src,
	alt,
	showRating,
	rating,
	children,
	onClick,
}: Props) {
	const lang = useParams()?.lang as Locales ?? fallbackLng;
	const { t } = useTranslation(lang);

	return (
		<Link
			className={`flex flex-col max-w-[10.625rem] group/card ${className}`}
			href={`${routes.persons}${personId}`}
			onClick={onClick}
			title={title}
			data-testid={testId}
		>
			<div className='w-[9.625rem] h-[9.625rem] relative self-center mb-3'>
				<ThemedImage
					className='rounded-[50%] object-cover w-full h-full'
					width={154}
					height={154}
					alt={alt}
					src={{
						light: src,
						dark: src,
					}}
					fallback={{
						light: '/assets/images/person-card-placeholder-l.svg',
						dark: '/assets/images/person-card-placeholder-d.svg'
					}}
				/>
				{showRating && (
					<span className='absolute bottom-0 left-0 w-10 h-10 overflow-hidden flex items-center justify-center rounded-[50%] border-2 border-blue-700 dark:border-blue-400 bg-neutral-200 dark:bg-dark-neutral-300 text-blue-700 dark:text-blue-300 font-bold'>
						<span className='sr-only'>{t('personCard.rating')}</span>
						{Math.trunc(rating ?? 0)}
					</span>
				)}
			</div>
			{title && (
				<Title
					className='mb-1 line-clamp-2 text-neutral-1000 dark:text-dark-neutral-900 group-hover/card:text-neutral-800 group-active/card:text-neutral-700 dark:group-hover/card:text-dark-neutral-700 dark:group-active/card:text-dark-neutral-600 transition-colors duration-150'
					as={titleElement}
					level={titleLevel}
				>
					{title}
				</Title>
			)}
			{children && (
				<p
					className='text-100 text-neutral-900 dark:text-dark-neutral-800 line-clamp-3'
				>
					{children}
				</p>
			)}
		</Link>
	);
}