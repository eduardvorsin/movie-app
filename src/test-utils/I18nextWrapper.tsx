import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;