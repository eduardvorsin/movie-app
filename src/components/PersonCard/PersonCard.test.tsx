import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import PersonCard from './PersonCard';
import i18next from '@/i18n/client';
import I18nextWrapper from '@/test-utils/I18nextWrapper';

describe('PersonCard tests', () => {
	it('is rendered correctly', async () => {
		const dictionary = { rating: 'Popularity' };

		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
			>
				test text
			</PersonCard>
		);

		expect(await screen.findByTestId<HTMLAnchorElement>('test card')).toBeInTheDocument();
	});

	it('when click on the card, the mock function should work', async () => {
		const mockFn = jest.fn();
		const user = userEvent.setup();
		const dictionary = { rating: i18next.t('personCard.rating') };

		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
				onClick={mockFn}
			>
				test text
			</PersonCard>
		);

		await screen.findByTestId<HTMLAnchorElement>('test card');

		await user.click(screen.getByRole<HTMLAnchorElement>('link'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('if title is not an empty string, then the title is rendered', async () => {
		const dictionary = { rating: 'Popularity' };

		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				title='test title'
				dictionary={dictionary}
			>
				test text
			</PersonCard>
		);
		await screen.findByTestId<HTMLAnchorElement>('test card');

		expect(screen.getByRole<HTMLHeadingElement>('heading')).toBeInTheDocument();
	});

	it('if showRating is true it should show the rating value', async () => {
		const dictionary = { rating: 'Popularity' };

		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				rating={30}
				showRating
				dictionary={dictionary}
			>
				test text
			</PersonCard>
		);
		await screen.findByTestId<HTMLAnchorElement>('test card');

		expect(screen.getByText<HTMLSpanElement>('30')).toBeInTheDocument();
	});

	it('is a basic snapshot', async () => {
		const dictionary = { rating: 'Popularity' };

		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
			>
				test text
			</PersonCard>
		);
		await screen.findByTestId<HTMLAnchorElement>('test card');

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});

	it('is a snapshot with title', async () => {
		const dictionary = { rating: 'Popularity' };

		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
			>
				test text
			</PersonCard>
		);
		await screen.findByTestId<HTMLAnchorElement>('test card');

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});

	it('is a snapshot with showRating', async () => {
		const dictionary = { rating: 'Popularity' };

		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
			>
				test text
			</PersonCard>
		);
		await screen.findByTestId<HTMLAnchorElement>('test card');

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});
});

describe('PersonCard integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('personCard.rating'));

		const dictionary = { rating: i18next.t('personCard.rating') };
		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
				rating={90}
				showRating
			>
				test text
			</PersonCard>,
			{ wrapper: I18nextWrapper }
		);

		await screen.findByTestId<HTMLAnchorElement>('test card');
		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('personCard.rating'));

		const dictionary = { rating: i18next.t('personCard.rating') };
		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
				rating={90}
				showRating
			>
				test text
			</PersonCard>,
			{ wrapper: I18nextWrapper }
		);

		await screen.findByTestId<HTMLAnchorElement>('test card');
		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');

		const dictionary = { rating: i18next.t('personCard.rating') };
		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
				rating={90}
				showRating
			>
				test text
			</PersonCard>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByTestId<HTMLAnchorElement>('test card')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');

		const dictionary = { rating: i18next.t('personCard.rating') };
		render(
			<PersonCard
				personId={1}
				src='/'
				alt='alternative text'
				testId='test card'
				dictionary={dictionary}
				rating={90}
				showRating
			>
				test text
			</PersonCard>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByTestId<HTMLAnchorElement>('test card')).toMatchSnapshot();
	});
});