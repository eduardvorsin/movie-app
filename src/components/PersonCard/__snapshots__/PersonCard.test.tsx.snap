// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonCard tests is a basic snapshot 1`] = `
<a
  class="flex flex-col max-w-[10.625rem] group/card undefined"
  data-testid="test card"
  href="/persons/1"
>
  <div
    class="relative self-center mb-3"
  >
    <img
      alt="alternative text"
      class="w-[9.625rem] h-[9.625rem] rounded-[50%] object-cover"
      data-nimg="1"
      decoding="async"
      height="231"
      loading="lazy"
      src="/_next/image?url=%2F&w=384&q=75"
      srcset="/_next/image?url=%2F&w=256&q=75 1x, /_next/image?url=%2F&w=384&q=75 2x"
      style="color: transparent;"
      width="154"
    />
  </div>
  <p
    class="text-100 text-neutral-900 dark:text-dark-neutral-800 line-clamp-3"
  >
    test text
  </p>
</a>
`;

exports[`PersonCard tests is a snapshot with showRating 1`] = `
<a
  class="flex flex-col max-w-[10.625rem] group/card undefined"
  data-testid="test card"
  href="/persons/1"
>
  <div
    class="relative self-center mb-3"
  >
    <img
      alt="alternative text"
      class="w-[9.625rem] h-[9.625rem] rounded-[50%] object-cover"
      data-nimg="1"
      decoding="async"
      height="231"
      loading="lazy"
      src="/_next/image?url=%2F&w=384&q=75"
      srcset="/_next/image?url=%2F&w=256&q=75 1x, /_next/image?url=%2F&w=384&q=75 2x"
      style="color: transparent;"
      width="154"
    />
    <span
      class="absolute bottom-0 left-0 w-10 h-10 overflow-hidden flex items-center justify-center rounded-[50%] border-2 border-blue-700 dark:border-blue-400 bg-neutral-200 dark:bg-dark-neutral-300 text-blue-700 dark:text-blue-300 font-bold"
    >
      <span
        class="sr-only"
      >
        personCard.rating
      </span>
      30
    </span>
  </div>
  <p
    class="text-100 text-neutral-900 dark:text-dark-neutral-800 line-clamp-3"
  >
    test text
  </p>
</a>
`;

exports[`PersonCard tests is a snapshot with title 1`] = `
<a
  class="flex flex-col max-w-[10.625rem] group/card undefined"
  data-testid="test card"
  href="/persons/1"
>
  <div
    class="relative self-center mb-3"
  >
    <img
      alt="alternative text"
      class="w-[9.625rem] h-[9.625rem] rounded-[50%] object-cover"
      data-nimg="1"
      decoding="async"
      height="231"
      loading="lazy"
      src="/_next/image?url=%2F&w=384&q=75"
      srcset="/_next/image?url=%2F&w=256&q=75 1x, /_next/image?url=%2F&w=384&q=75 2x"
      style="color: transparent;"
      width="154"
    />
  </div>
  <h6
    class="-tracking-[0.01em] text-300 sm:text-400  leading-3 font-bold mb-1 line-clamp-2 text-neutral-1000 dark:text-dark-neutral-900 group-hover/card:text-neutral-800 group-active/card:text-neutral-700 dark:group-hover/card:text-dark-neutral-700 dark:group-active/card:text-dark-neutral-600 transition-colors duration-150"
  >
    test title
  </h6>
  <p
    class="text-100 text-neutral-900 dark:text-dark-neutral-800 line-clamp-3"
  >
    test text
  </p>
</a>
`;
